state/protocol/protocol_state/updater_test.go

```diff
@@ -0,0 +1,464 @@
+		expectedCurrentEpochIdentityTable = append(expectedCurrentEpochIdentityTable, &flow.DynamicIdentityEntry{
+			NodeID: participant.NodeID,
+			Dynamic: flow.DynamicIdentity{
+				Weight:  participant.Weight,
+				Ejected: participant.Ejected,
+			},
+		})
+
+		expectedNextEpochIdentityTable = append(expectedNextEpochIdentityTable, &flow.DynamicIdentityEntry{
+			NodeID: participant.NodeID,
+			Dynamic: flow.DynamicIdentity{
+				Weight:  0,
+				Ejected: participant.Ejected,
+			},
+		})
+	}
+
+	// in this loop we perform a few extra lookups to avoid duplicates
+	for _, participant := range setup.Participants {
+		if _, found := expectedCurrentEpochIdentityTable.ByNodeID(participant.NodeID); !found {
+			expectedCurrentEpochIdentityTable = append(expectedCurrentEpochIdentityTable, &flow.DynamicIdentityEntry{
+				NodeID: participant.NodeID,
+				Dynamic: flow.DynamicIdentity{
+					Weight:  0,
+					Ejected: participant.Ejected,
+				},
+			})
+		}
+
+		entry := &flow.DynamicIdentityEntry{
+			NodeID: participant.NodeID,
+			Dynamic: flow.DynamicIdentity{
+				Weight:  participant.InitialWeight,
+				Ejected: participant.Ejected,
+			},
+		}
+		existing, found := expectedNextEpochIdentityTable.ByNodeID(participant.NodeID)
+		if found {
+			existing.Dynamic = entry.Dynamic
+		} else {
+			expectedNextEpochIdentityTable = append(expectedNextEpochIdentityTable, entry)
+		}
+	}
+	// finally, sort in canonical order
+	expectedCurrentEpochIdentityTable = expectedCurrentEpochIdentityTable.Sort(order.IdentifierCanonical)
+	expectedNextEpochIdentityTable = expectedNextEpochIdentityTable.Sort(order.IdentifierCanonical)
+
+	// process actual event
+	err := s.updater.ProcessEpochSetup(setup)
+	require.NoError(s.T(), err)
+
+	updatedState, _, hasChanges := s.updater.Build()
+	require.True(s.T(), hasChanges, "should have changes")
+	require.Equal(s.T(), expectedCurrentEpochIdentityTable, updatedState.Identities)
+	nextEpochProtocolState := updatedState.NextEpochProtocolState
+	require.NotNil(s.T(), nextEpochProtocolState, "should have next epoch protocol state")
+	require.Equal(s.T(), nextEpochProtocolState.CurrentEpochEventIDs.SetupID, setup.ID(),
+		"should have correct setup ID for next protocol state")
+	require.False(s.T(), nextEpochProtocolState.InvalidStateTransitionAttempted, "should not set invalid state transition attempted")
+	require.Nil(s.T(), nextEpochProtocolState.NextEpochProtocolState, "should not have next epoch protocol state")
+	require.Equal(s.T(), expectedNextEpochIdentityTable, nextEpochProtocolState.Identities)
+}
+
+// TestProcessEpochSetupWithSameParticipants tests that processing epoch setup with overlapping participants results in correctly
+// built updated protocol state. It should build a union of participants from current and next epoch for current and
+// next epoch protocol states respectively.
+func (s *UpdaterSuite) TestProcessEpochSetupWithSameParticipants() {
```

Comment:
I find that just relying on length is a bit dangerous. I would suggest to extend this test (similarly to the one before) to check identities explicitly

Reply:
https://github.com/onflow/flow-go/pull/4579/commits/d222fd0a645ebd03a8c27df26857cee47c354689